### DISCUSSION TITLE
Rule exclusion for a particular resource (Terraform)

### DISCUSSION THREAD

**Discussion by wplj** [aquasecurity/trivy#5367] [Q&A]
### Question

Is it possible? For example, I want to exempt one storage account from storage firewall check (public access), but keep this rule active for all other accounts.

### Target

Filesystem

### Scanner

Vulnerability

### Output Format

Table

### Mode

Standalone

### Operating System

Windows 10

### Version

```bash
Version: 0.45.1                                                                  
Vulnerability DB:                                                                
  Version: 2                                                                     
  UpdatedAt: 2023-10-11 06:23:24.848988623 +0000 UTC                             
  NextUpdate: 2023-10-11 12:23:24.848988123 +0000 UTC                            
  DownloadedAt: 2023-10-11 10:28:27.4985531 +0000 UTC                            
Policy Bundle:                                                                   
  Digest: sha256:1df8ade71efc830877ca3b1130f83e0c6368e3a45b0d4c0f0418955501644054
  DownloadedAt: 2023-10-11 10:28:28.0684779 +0000 UTC
```

**Comment by nikpivkin** [comment]
Hi @wplj !

Do you mean exclude a rule for a resource by name?

**Reply by wplj** [reply]
yes, by name (or Terraform id, e.g., `azurerm_storage_account.tfstate`).

**Reply by simar7** [reply]
Can you share a sample terraform config to show us how you would use this?

**Reply by wplj** [reply]
`tfsec` can read special comments in HCL files, like
`#tfsec:ignore:azure-storage-default-action-deny` on the lines that cause check failures, and filter them accordingly.
I was thinking about something similar but declared in a separate file (`.trivyignore`?) like this (key-value pairs, basically):
```
terraform_ignore:
  azurerm_network_security_group.nsg-trusted: azure-network-no-public-egress, azure-network-no-public-ingress
  azurerm_storage_account.tfstate: azure-storage-default-action-deny
```
etc.

**Reply by simar7** [reply]
I see. Thanks for the idea but at the moment we don't have plans to add support for that.

**Reply by itaysk** [reply]
@simar7 wouldn't they be able to use [filtering with rego](https://aquasecurity.github.io/trivy/v0.46/docs/configuration/filtering/#by-open-policy-agent) for that?

**Reply by simar7** [reply]
@wplj we have support for ignoring per file (for each misconf) in `.trivyignore` as such:

```shell
misconfigurations:
  - id: AVD-DS-0001
  - id: AVD-DS-0002
    paths:
      - "docs/Dockerfile"
    statement: The image needs root privileges
```

Will ignore the two misconfigurations but will only ignore `AVD-DS-0002` in the `docs/Dockerfile` path only. 

As for what @itaysk has suggested, you can also pass in an ignore policy as such (support for this will be landing when https://github.com/aquasecurity/trivy/pull/5359 is merged. But you can see an example below of how to use, once it is available.

```shell
trivy config --ignore-policy=<path/to/policy.rego> <iac source code>
```

Where `policy.rego` could look something as follows:

```rego
package trivy

import data.lib.trivy

default ignore=false

ignore {
	input.CauseMetadata.Resource == "azure-network-no-public-egress"
}
```

Will ignore the results in `azure-network-no-public-egress`.

As for which fields you can filter on, here's a list of supported input types in the case of misconfiguration scanning https://github.com/aquasecurity/trivy/blob/af89249deab0a9e4ed1a959cedf8b5ace731b45a/pkg/types/misconfiguration.go#L7-L21

I hope this helps. Please let us know if you need assistance.

**Reply by itaysk** [reply]
> you can also pass in an ignore policy as such (support for this will be landing when https://github.com/aquasecurity/trivy/pull/5359 is merged

you don't need to wait for that actually, this already works with `trivy fs`
